doc/developer/design/20241028_optimizer_releases.md

```diff
@@ -0,0 +1,81 @@
+
+## Minimal Viable Prototype
+
+### Version names and numbering
+
+Every version of the `optimizer` crate will be numbered. Customers can run SQL like:
+
+```sql
+ALTER CLUSTER name SET (OPTIMIZER VERSION = version)
+```
+
+where `version` is a version number (or, possibly one of our three fixed names). Note that the optimizer is selected on a per cluster basis. Selecting the optimizer per cluster makes it easier to qualify optimizer releases.
+
+### The `optimizer` and `optimizer-types` crates
+
+The `optimizer` crate will contain the definitions of HIR, MIR, and LIR, along with the HIR-to-MIR and MIR-to_LIR lowerings and the MIR-to-MIR transformations. These come from the `sql` (HIR, HIR -> MIR), `expr` (MIR), `transform` (MIR -> MIR), and `compute-types` (LIR, MIR -> LIR) crates.
+
+The `optimizer-types` crate will have traits and definitions that are global across all three live versions of the optimizer. The AST types may change version to version, but the traits will be more stable.
+
+These crates do _not_ include:
+  + the SQL parser
+  + (\*) SQL -> HIR in `sql`
+  + `sequence_*` methods in adapter
+  + (\*) LIR -> dataflow in `compute`
+
+The two bullets marked (\*) above are a tricky point in the interface. SQL will _always_ lowers to **unstable**/**hot** HIR; we will have methods to convert **unstable**/**hot** HIR to the other two versions. Similarly, LIR lowers to dataflow from **unstable**/**hot** LIR; we will have methods to convert the other two versions to the latest.
+
+### Supporting qualification
+
+A capture-and-replay mode for events that stores the events in the customer's own environment would allow for (a) customers to easily test and qualify changes, and (b) a push-button way for us to spin up unbilled clusters to evaluate these replays in read-only mode.
+
+## Alternatives
+
+The thorniest thing here is the release cutting and copied code. Would it be better to split off the optimizer entirely into a separate repository? A separate _process_?
+
+## Open questions
+
+- Does **unstable**/**hot** get a version number, or no? (Cf. Debian sid)
+
+- How precisely do we select features out of **unstable**/**hot** to cut a new **testing**/**warm**? If we literally copy the crate three times in the repo, `git cherry-pick` will not be straightfoward to use and it will be easy to lose history.
+
+- Under use-case isolation, can we say select an optimizer per `environmentd` instead of per cluster?
+
+- There are several occasions where the optimizer is called independent of the cluster: views, prepared statements, and as part of `INSERT`, `UPDATE`, and `DELETE`. (Look at `sequence_read_then_write` for an example.) Should these be moved off? Use the active cluster? Automatically optimize with all three?
+
+- Which fixes will we backport? Correctness only? When do we backport a fix and when do we backport a "problem detector" with a notice?
```

Comment:
Do we need to consider new features in addition to bugfix backports? If we introduce some new SQL-level feature that requires optimizer support (like WMR), all three versions of the optimizer need to support that, right? Or is the idea that adapter will take the optimizer version into account when planning a user query and reject the query if the optimizer doesn't support it?

Reply:
> Or is the idea that adapter will take the optimizer version into account when planning a user query and reject the query if the optimizer doesn't support it?

This, yeah, although it won't be the adapter that checks the version. Here's my understanding. Adapter (i.e., the `sql`/`adapter` crates) always produces an HIR plan for the latest version of the optimizer. Each optimizer supports converting its HIR plan to the previous version's HIR plan. This conversion can fail, if the plan uses features that are not available in the previous optimizer. E.g., if there's a new HIR node in hot that's not in warm, if the user has requested the warm optimizer, the query will be rejected when trying to convert from a hot to warm plan.

There may occasionally be moments where implementing the version check in the HIR -> HIR conversion inside the optimizer is too painful, and we need to hoist the version check into the adapter, but ideally we can avoid that whenever possible.

Reply:
Cool, that makes sense to me. We'll want to disallow changing the optimizer of an existing cluster, or at least disallow changing it into the colder direction. Otherwise we could end up with objects installed on a cluster that cannot be handled by that cluster's optimizer anymore during, e.g., a version  upgrade.